src/Discord.Net.WebSocket/Entities/Interaction/Message Components/SocketMessageComponent.cs

```diff
@@ -109,45 +109,104 @@ internal override void Update(Model model)
+            var args = new MessageProperties();
+            func(args);
+
+            if (!IsValidToken)
+                throw new InvalidOperationException("Interaction token is no longer valid");
+
+            if (args.AllowedMentions.IsSpecified)
+            {
+               var allowedMentions = args.AllowedMentions.Value;
+               Preconditions.AtMost(allowedMentions?.RoleIds?.Count ?? 0, 100, nameof(allowedMentions), "A max of 100 role Ids are allowed.");
+               Preconditions.AtMost(allowedMentions?.UserIds?.Count ?? 0, 100, nameof(allowedMentions), "A max of 100 user Ids are allowed.");
+            }
+
+            if (args.Embeds.IsSpecified)
+                Preconditions.AtMost(args.Embeds.Value?.Length ?? 0, 10, nameof(args.Embeds), "A max of 10 embeds are allowed.");
+
+            // check that user flag and user Id list are exclusive, same with role flag and role Id list
+            if (args.AllowedMentions.IsSpecified && args.AllowedMentions.Value != null && args.AllowedMentions.Value.AllowedTypes.HasValue)
+            {
+                var allowedMentions = args.AllowedMentions.Value;
+                if (allowedMentions.AllowedTypes.Value.HasFlag(AllowedMentionTypes.Users)
+                && allowedMentions.UserIds != null && allowedMentions.UserIds.Count > 0)
+                {
+                    throw new ArgumentException("The Users flag is mutually exclusive with the list of User Ids.", nameof(args.AllowedMentions));
+                }
 
+                if (allowedMentions.AllowedTypes.Value.HasFlag(AllowedMentionTypes.Roles)
+                && allowedMentions.RoleIds != null && allowedMentions.RoleIds.Count > 0)
+                {
+                    throw new ArgumentException("The Roles flag is mutually exclusive with the list of Role Ids.", nameof(args.AllowedMentions));
+                }
+            }
+
+            var response = new API.InteractionResponse
+            {
+                Type = InteractionResponseType.UpdateMessage,
+                Data = new API.InteractionCallbackData
+                {
+                    Content = args.Content,
+                    AllowedMentions = args.AllowedMentions.IsSpecified ? args.AllowedMentions.Value?.ToModel() : Optional<API.AllowedMentions>.Unspecified,
+                    Embeds = args.Embeds.IsSpecified ? args.Embeds.Value?.Select(x => x.ToModel()).ToArray() : Optional<API.Embed[]>.Unspecified,
+                    Components = args.Components.IsSpecified
+                        ? args.Components.Value?.Components.Select(x => new API.ActionRowComponent(x)).ToArray()
+                        : Optional<API.ActionRowComponent[]>.Unspecified,
+                    Flags = args.Flags.IsSpecified ? (int?)args.Flags.Value ?? Optional<int>.Unspecified : Optional<int>.Unspecified
+                }
+            };
+
+            await InteractionHelper.SendInteractionResponse(this.Discord, response, this.Id, this.Token, options);
```

Comment:
Discord doesnt upsert interaction responses, have a look at https://discord.com/developers/docs/interactions/slash-commands#edit-original-interaction-response

Reply:
You could probably move this to the abstract `SocketInteraction` class.

Reply:
I don't quite understand what you're talking about :\

Reply:
Your not modifying/updating the interaction response. your just sending another one.

Reply:
Yes

Reply:
This method updates the message of the component on which the interaction was received on.

Reply:
For modify original response, i was add method `ModifyOriginalResponseAsync`

Reply:
I still dont understand this method, could you explain what it does fully?

Reply:
When you receive component interaction, you can use `UpdateAsync` for edit message with it component and response to this interaction, and it all in one request. Its just response with type `UpdateMessage`.

Reply:
But it work more fine than old `RespondAsync` with type `UpdateMessage`

Reply:
It also more corresponds to discord.net

Reply:
Can you send a snipped how how it would be used

Reply:
```cs
var component = new ComponentBuilder().WithButton(label: "click", customId: "test");
await Context.Channel.SendMessageAsync(text: "Waiting for click...", component: component.Build()).ConfigureAwait(false);

client.InteractionCreated += Client_InteractionCreated;
int clickCount = 0;
private async Task Client_InteractionCreated(SocketInteraction interaction)
{
    if (interaction is not SocketMessageComponent { Data: { CustomId: "test" } } componentInteraction)
        return;
     
    await componentInteraction.UpdateAsync(props => props.Content = $"Clicked `{clickCount++}!`").ConfigureAwait(false);
}
```
![test](https://user-images.githubusercontent.com/47792796/127124978-bd3677eb-b81c-4da3-93a1-234c5168f52d.gif)